dpypelines/pipeline/shared/transforms/sdmx/generic/v21/prototype/v1.py

```diff
@@ -0,0 +1,302 @@
+    get_number_of_obs_from_xml_file,
+)
+
+
+def xmlToCsvSDMX2_1(input_path, output_path):
+
+    check_read_in_sdmx(input_path)  # transform validation
+
+    # Converting the XML file into a giant dictionary from the nested header dictionary
+    with open(input_path, "r") as file:
+        xml_content = file.read()
+        data = xmltodict.parse(xml_content)
+        check_xml_type(data)  # transform validation
+
+    header = data["message:GenericData"]["message:Header"]
+    for key in header.keys():
+        header[(key.replace("message:", "").replace("common:", ""))] = header.pop(key)
+    check_header_info(header)  # transform validatio
+
+    expected_number_of_obs = get_number_of_obs_from_xml_file(
+        input_path
+    )  # transform validation
+
+    tables = data["message:GenericData"]["message:DataSet"]["generic:Series"]
+
+    series_dict = {}
+    obs_dicts = []
+
+    for table in tables:
+        for element in table.items():
+            if element[0] == "generic:SeriesKey":
+                df = pd.DataFrame(element[1]["generic:Value"])
+                series_dict = {a: b for a, b in zip(df["@id"], df["@value"])}
+            elif element[0] == "generic:Attributes":
+                df = pd.DataFrame(element[1]["generic:Value"])
+                series_dict.update({a: b for a, b in zip(df["@id"], df["@value"])})
+            elif element[0] == "generic:Obs":
+                for observation in element[1]:
+                    observation_temp = {}
+                    df = pd.DataFrame(
+                        observation["generic:Attributes"]["generic:Value"]
+                    )
+                    observation_temp.update(
+                        {a: b for a, b in zip(df["@id"], df["@value"])}
+                    )
+                    observation_temp.update(
+                        {
+                            "generic:ObsDimension": next(
+                                iter(observation["generic:ObsDimension"].values())
+                            )
+                        }
+                    )
+                    observation_temp.update(
+                        {
+                            "generic:ObsValue": next(
+                                iter(observation["generic:ObsValue"].values())
+                            )
+                        }
+                    )
+                    obs_dicts.append(series_dict | observation_temp)
+            else:
+                print("something fucked up")
+
+    obs_frame = pd.DataFrame(obs_dicts)
+    check_length_of_dataframe_is_expected_length(
+        obs_frame, expected_number_of_obs
+    )  # transform validation
+
+    header_dict = flatten_dict(header)
+    check_header_unpacked(header_dict)  # transform validation
+    header_df = pd.DataFrame([header_dict])
+
+    repl_rows = header_df.loc[0].copy()
+    header_frame = pd.concat(
+        [header_df, pd.DataFrame([repl_rows] * (len(obs_frame) - 1))], ignore_index=True
+    )
```

Comment:
I've found that 
```
pd.DataFrame([repl_rows] * (len(obs_frame) - 1))]
```
is also quite inefficient. I have a suggestion to replace it with, it may seem pedantic but I'm just concerned that for much larger datasets it will slow it down unnecessarily. My suggestion is to replace the above 4 lines with
```
header_dict_2 = header_df.to_dict(orient='list')
for key in header_dict_2:
        header_dict_2[key] = header_dict_2[key] * (len(obs_frame) - 1)
    
 header_frame_2 = pd.DataFrame(header_dict_2)
 header_frame = pd.concat(
        [header_df, header_frame_2], ignore_index=True
    )
```
essentially you are using a dict and duplicating items in a list to the required length, rather than duplicated a pd.Series (again its pandas that is inefficient)